api/v1beta1/glanceapi_webhook.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// GlanceAPIDefaults -
+type GlanceAPIDefaults struct {
+	ContainerImageURL string
+}
+
+var glanceAPIDefaults GlanceAPIDefaults
+
+// log is for logging in this package.
+var glanceapilog = logf.Log.WithName("glanceapi-resource")
+
+// SetupWebhookWithManager sets up the webhook with the Manager
+func (r *GlanceAPI) SetupWebhookWithManager(mgr ctrl.Manager, defaults GlanceAPIDefaults) error {
+	glanceAPIDefaults = defaults
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-glance-openstack-org-v1beta1-glanceapi,mutating=true,failurePolicy=fail,sideEffects=None,groups=glance.openstack.org,resources=glanceapis,verbs=create;update,versions=v1beta1,name=mglanceapi.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &GlanceAPI{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *GlanceAPI) Default() {
+	glanceapilog.Info("default", "name", r.Name)
+
+	r.Spec.Default()
+}
+
+// Default - set defaults for this GlanceAPI spec
+func (spec *GlanceAPISpec) Default() {
+	if spec.ContainerImage == "" {
+		spec.ContainerImage = glanceAPIDefaults.ContainerImageURL
```

Comment:
This feels redundant now. The `func (spec *GlanceSpec) Default()` on Glance level already set a value to this field if it was empty so this code will never be triggered. 

I see couple of possible solutions alternatives:

1. `func (spec *GlanceSpec) Default()` is changed to call `Glance.Spec.GlanceAPIExternal.Default()` and `Glance.Spec.GlanceAPIInternal.Default()` respectively. In this case the webhook registration is not needed in GlanceAPI at all. This also means that openstack-operator only need to call `func (spec *GlanceSpec) Default()`  and the rest is automatic. However the downside of this is that if GlanceAPI is ever instantiated independently from Glance struct then the defaulting on GlanceAPI will not be triggered.

2. `func (spec *GlanceSpec) Default()` can simply not default anything on any sub Spec struct and let the webhook in GlanceAPI level do it.  This means that openstack-operator needs to call `func (spec *GlanceSpec) Default()`  and `func (spec *GlanceAPISpec) Default()` as well.

Reply:
This is a key point of this implementation: I think we don't have to expose at `openstack-operator` level the subCRs specs, and from that layer we should call `func (spec *GlanceSpec) Default()`.
It's the service implementation that then triggers  `Glance.Spec.GlanceAPIExternal.Default()` and `Glance.Spec.GlanceAPIInternal.Default()` to enforce defaults (it's the same in case of `ContainerImage` but it could be different for other parameters).
So I think option 1. could be a good choice in a top-down view, but I see there's a problem on the `GlanceAPI` trigger.
@abays wdyt about the two options?

Reply:
> However the downside of this is that if GlanceAPI is ever instantiated independently from Glance struct then the defaulting on GlanceAPI will not be triggered.

That is exactly why I had initially included the redundant `GlanceAPI` defaults.  I don't know where we are right now on whether we'll support individual, direct sub-resource creation as a valid user workflow.  I think we should come to a consensus in that regard and then that decision will inform the proper approach to these defaulting patterns.

Reply:
Let's try to discuss this tomorrow on the sync meeting.

Reply:
We discussed the sub-resource CR concern today, and I believe the consensus is that we won't support creating/applying them.  That is, in the context of this particular operator, only `Glance` CRs should ever be created.  So I think we can remove the `GlanceAPI` defaulting logic.  @gibizer was that your understanding from today's conversation as well?

Reply:
Yeah, so we only add the Defaulting interface to the Glance level but never on GlanceAPI level. And the Glance level defaulter will do defaulting on the GlanceAPISpec stuct as well.

Reply:
+1 on that